### DISCUSSION TITLE
PWM Speed Limit

### DISCUSSION THREAD

**Discussion by unixb0y** [arendst/Tasmota#22292] [Q&A] [answered]
First of all: I'm trying to drive a 110kHz square wave using the D1 mini.

However, when configuring a GPIO to be PWM output, the frequency on ESP8266 can only be set up to 4kHz, meanwhile the IC should support higher output speeds, e.g. when clocking the SPI CLK line.

People online mention being able to do SPI on the ESP8266 at several MHz; could this 4kHz limit be increased in Tasmota or would it cause issues with all the other stuff that has to be run (web UI, ....)?

**Comment by Jason2866** [comment]
Not possible with Tasmota with an esp8266.

**Comment by s-hadinger** [comment]
PMW is done in software via interrupts on ESP8266, hence the frequency limit. ESP32 is hardware, so you can go much higher.

It is certainly possible to do some trickery with SPI to output arbitrary high frequencies, but this is not supported in Tasmota. Best option is using ESP32 Wemos D1 mini

**Reply by unixb0y** [reply]
I'm still not clear why ESP8266 can do 80MHz software SPI in theory, but when used in Tasmota, the same chip cannot output more than a couple kHz? That's a 3 order of magnitude difference. 🤔 

Since Tasmota does support some SPI-based peripherals, what would be the max. SPI speed for those (say an SPI screen)?

**Reply by s-hadinger** [reply]
Well Tasmota uses SPI for SPI communication, no more no less. PWM is done in software and does not use SPI hardware.

What you are asking for would require someone to develop a specific driver to use SPI controller for non-SPI usage. Currently what you are asking for is not supported in Tasmota.

**Reply by unixb0y** [reply]
So you're saying if I develop a driver that gives front-end access to the SPI clock, that would allow me to create a 100kHz square wave?

**Reply by Jason2866** [reply]
Maybe? Technical reference https://www.mpja.com/download/esp8266%20technical%20reference.pdf
Easier to do with a esp32 sind there are more hardware units (pwm, ledc, RMT)  where this frequencys can be reached. Anyways not the goal of Tasmota to be a sort of frequency generator

**Reply by sfromis** [reply]
When you get 80 MHz SPI with ESP8266, it is not software SPI, but using the feature of hardware SPI.

Also, PWM is more complex than a simple clock, due to variable pulse width, and synchronizing multiple PWM channels with separate duty cycles.

**Reply by unixb0y** [reply]
That's good to know; so the ESP8266 does have hardware SPI as I suspected. I don't really need PWM, just a high-frequency square wave that I can turn on / off and configure its frequency. I'll try to implement that as a driver, then.

**Comment by lalo-uy** [comment]
Don’t forget that the SPI on EsP8266 is used to access the flash memory.

**Reply by unixb0y** [reply]
Sure; but it's also used to access SPI displays and other SPI peripherals that are supported by Tasmota, so I don't see why my proposed driver shouldn't work?

**Reply by sfromis** [reply, marked as answer]
AFAIK, you do not get to choose some SPI frequency, but instead select a clock divider, basically how many time to divide the CPU frequency by two, like:
`SPI.setClockDivider(SPI_CLOCK_DIV8);`

Using ESP32 PWM you can get the flexibility you seem to want.

**Reply by unixb0y** [reply]
Alright, thanks, then I'll try with an ESP32-C3! Have some spare C3 minis.
